File: DesignPatterns/01_GOF23/DP_examples/Python/02_Structural/Facade/Facade.py
from Subsystem import Circle, Rectangle, Square, IShape

# 形状类型枚举
class ShapeType:
    CIRCLE = 0
    RECTANGLE = 1
    SQUARE = 2


def build_shape(shape_type, *args) -> IShape:
    if shape_type == ShapeType.CIRCLE:
        if len(args) >= 1:
            return Circle(args[0])
    elif shape_type == ShapeType.RECTANGLE:
        if len(args) >= 2:
            return Rectangle(args[0], args[1])
    elif shape_type == ShapeType.SQUARE:
        if len(args) >= 1:
            return Square(args[0])
    return None

File: DesignPatterns/01_GOF23/DP_examples/Go/02_Structural/Facade/facade.go
package Facade

// 形状类型枚举
type ShapeType int

const (
	CIRCLE ShapeType = iota
	RECTANGLE
	SQUARE
)

// 形状创建者（外观类）
type shapeMaker struct{}

// 单例实例
var ShapeMaker shapeMaker

// 构建形状
func (sm shapeMaker) BuildShape(shapeType ShapeType, args ...int) IShape {
	switch shapeType {
	case CIRCLE:
		if len(args) >= 1 {
			return NewCircle(args[0])
		}
	case RECTANGLE:
		if len(args) >= 2 {
			return NewRectangle(args[0], args[1])
		}
	case SQUARE:
		if len(args) >= 1 {
			return NewSquare(args[0])
		}
	}
	return nil
}
